[DOC] Instructions for oauth 2.0 token based authentication (#2013)
strimzi/strimzi-kafka-operator: documentation/book/oauth/proc-oauth-broker-config.adoc

// Module included in the following module:
//
// con-oauth-config.adoc

[id='proc-oauth-broker-config-{context}']
= Configuring {oauth} support for Kafka brokers

This procedure describes how to configure Kafka brokers so that the broker listeners are enabled to use {oauth} authentication using an authorization server.

If the authorization server is using certificates signed by trusted CA and matching the {oauth} server hostname, TLS connection works out of box.
Otherwise, you have two connection options for your external listener configuration:

. TLS connection to the authorization server with trusted certificates
. Direct connection using an introspection endpoint configuration

Both options are described in this procedure.

.Before you start

For more information on the configuration and authentication of Kafka broker listeners, see:

* xref:assembly-configuring-kafka-broker-listeners-deployment-configuration-kafka[Kafka broker listeners]
* xref:assembly-kafka-authentication-and-authorization-deployment-configuration-kafka[Authentication and Authorization]

.Prerequisites

* {ProductName} and Kafka is running
* {oauth} authorization server is deployed

.Procedure

> . Update the Kafka broker configuration (`Kafka.spec.kafka`) of your `kafka` resource in an editor.

Reviewer: it should be with capital K, so `Kafka`

+
[source,shell]
----
kubectl edit kafka my-cluster
----

> . Configure the Kafka broker `listeners` configuration

Reviewer 2: Missing full stop.

// Module included in the following assemblies:
//
// assembly-oauth-strimzi-config.adoc

[id='proc-oauth-broker-config-{context}']
= Configuring {oauth} support for Kafka brokers

This procedure describes how to configure Kafka brokers so that the broker listeners are enabled to use {oauth} authentication using an authorization server.

> In this scenario, {ProductName} and the authorization server are running in the same cluster.

Reviewer 3: That is actually not necessarily true. I think you can remove this.

> If the authorization server is using self-signed certificates with valid hostnames, TLS connection works out of box.

Reviewer 3: ```suggestion
If the authorization server is using certificates signed by trusted CA and matching the {oauth} server hostname, TLS connection works out of box.
```

Otherwise, you have two connection options for your external listener configuration:

. TLS connection to the authorization server with trusted certificates
. Direct connection using an introspection endpoint configuration

Both options are described in this procedure.

.Before you start

For more information on the configuration and authentication of Kafka broker listeners, see:

* xref:assembly-configuring-kafka-broker-listeners-deployment-configuration-kafka[Kafka broker listeners]
* xref:assembly-kafka-authentication-and-authorization-deployment-configuration-kafka[Authentication and Authorization]

.Prerequisites

* {ProductName} and Kafka is running
* {oauth} authorization server is deployed

.Procedure

. Update the Kafka broker configuration (`Kafka.spec.kafka`) of your `kafka` resource in an editor.
+
[source,shell]
----
kubectl edit kafka my-cluster
----

> . Configure the Kafka broker listener configuration

Reviewer 3: Could you please clarify that the configuration is independent for each listener? It doesn't have to be the same.

+
For example:
+
[source,yaml,subs="+quotes,attributes"]
----
listeners:
 plain:
   authentication:
     type: oauth <1>

>      validIssuerUri: http://<authorization-server-address>.nip.io/auth/realms/demo <2>

Reviewer 3: ```suggestion
     validIssuerUri: https://<authorization-server-address>/auth/realms/demo <2>
```

>      jwksEndpointUri: http://<authorization-server-address>.nip.io/auth/realms/demo/protocol/openid-connect/certs <3>

Reviewer 3: ```suggestion
     jwksEndpointUri: https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/certs <3>
```

     userNameClaim: preferred_username <4>
 tls:
   authentication:
     type: oauth

>      validIssuerUri: http://<authorization-server-address>.nip.io/auth/realms/demo

Reviewer 3: ```suggestion
     validIssuerUri: https://<authorization-server-address>/auth/realms/demo
```

>      jwksEndpointUri: http://<authorization-server-address>.nip.io/auth/realms/demo/protocol/openid-connect/certs

Reviewer 3: ```suggestion
     jwksEndpointUri: https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/certs
```

     userNameClaim: preferred_username
 external:
   type: {external-type}
   //Option 1
   authentication: <5>
     type: oauth

>      validIssuerUri: http://<authorization-server-address>.nip.io/auth/realms/external

Reviewer 3: ```suggestion
     validIssuerUri: https://<authorization-server-address>/auth/realms/external
```

     jwksEndpointUri: http://<authorization-server-address>.nip.io/auth/realms/demo/protocol/openid-connect/certs
     userNameClaim: preferred_username
     tlsTrustedCertificates: <6>
     - secretName: oauth-server-cert
       certificate: ca.crt
   disableTlsHostnameVerification: true <7>

   //Option 2
   authentication: <8>
     type: oauth

>      validIssuerUri: https://sso.myproject.svc:8443/auth/realms/internal

Reviewer 3: ```suggestion
     validIssuerUri: https://<authorization-server-address>/auth/realms/external
```

>      introspectionEndpointUri: https://sso.myproject.svc:8443/auth/realms/internal/protocol/openid-connect/token/introspect

Reviewer 3: ```suggestion
     introspectionEndpointUri: https://<authorization-server-address>/auth/realms/external/protocol/openid-connect/token/introspect
```

     clientId: kafka-broker
     clientSecret:
       secretName: my-cluster-oauth
       key: clientSecret

----
<1> Listener type set to `oauth`.
<2> URI of the token issuer used for authentication.
<3> URI of the JWKS certificate endpoint used for local JWT validation.
<4> The user name profile claim (or key) that contains the actual user name in the token. The user name is the _principal_ used to identify the user. The `userNameClaim` value will depend on the authentication flow and the authorization server used.
<5> OPTION 1: TLS connection to the authorization server.
<6> (Optional) Trusted certificates for TLS connection to the authorization server.
<7> Enable TLS hostname verification. Default is `false`.
<8> OPTION 2: Introspection endpoint to connect directly to the authorization server.

. Save and exit the editor, then wait for rolling updates to complete.

. Check the update in the logs or by watching the pod state transitions:
+
[source,shell,subs="+quotes,attributes"]
----

> kubectl logs -f ${POD_NAME} ${CONTAINER_NAME}

Reviewer: it should be `-c ${CONTAINER_NAME}`

> kubectl get po -w

Reviewer: `pod` instead of `po`

     userNameClaim: preferred_username <4>
 tls:
   authentication:
     type: oauth
     validIssuerUri: _https://<authorization-server-address>/auth/realms/demo_
     jwksEndpointUri: _https://<authorization-server-address>/auth/realms/demo/protocol/openid-connect/certs_
     userNameClaim: preferred_username
 external:
   type: loadbalancer

   //Option 1
   authentication: <5>
     type: oauth
     validIssuerUri: _https://<authorization-server-address>/auth/realms/external_
     jwksEndpointUri: _https://<authorization-server-address>/auth/realms/external/protocol/openid-connect/certs_
     userNameClaim: preferred_username
     tlsTrustedCertificates: <6>
     - secretName: oauth-server-cert
       certificate: ca.crt
   disableTlsHostnameVerification: true <7>

   //Option 2
   authentication: <8>
     type: oauth
     validIssuerUri: _https://<authorization-server-address>/auth/realms/external_
     introspectionEndpointUri: _https://<authorization-server-address>/auth/realms/external/protocol/openid-connect/token/introspect_
     clientId: kafka-broker
     clientSecret:
       secretName: my-cluster-oauth

>        key: clientSecret

Author: I can drop the plain and tls. We want to distinguish the two external options though, right?

----
<1> Listener type set to `oauth`.
<2> URI of the token issuer used for authentication.
<3> URI of the JWKS certificate endpoint used for local JWT validation.
<4> The user name profile claim (or key) that contains the actual user name in the token. The user name is the _principal_ used to identify the user. The `userNameClaim` value will depend on the authentication flow and the authorization server used.
<5> OPTION 1: TLS connection to the authorization server.
<6> (Optional) Trusted certificates for TLS connection to the authorization server.
<7> Enable TLS hostname verification. Default is `false`.
<8> OPTION 2: Introspection endpoint to connect directly to the authorization server.

. Save and exit the editor, then wait for rolling updates to complete.

. Check the update in the logs or by watching the pod state transitions:
+
[source,shell,subs="+quotes,attributes"]
----
kubectl logs -f ${POD_NAME} -c ${CONTAINER_NAME}
kubectl get po -w
----
+

> The rolling updates configure the brokers to send messages using {oauth} authentication.

Reviewer 3: The brokers do more than send messages. Also, there should be just one rolling update. So maybe this would be better:

```suggestion
The rolling update configures the brokers to use {oauth} authentication.
```
